Add dynamic chart margin calculation and improve Y-axis tick formatting

chihsuan opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes CHARTS-2

## Proposed changes:
<!--- Explain what functional changes your PR includes -->


The line and bar charts use a fixed left margin of 40 to avoid Y-axis label overflow, but this doesn't work well when labels are wide, causing them to be cut off.



<img width="832" alt="Screenshot 2025-05-29 at 15 20 24" src="https://github.com/user-attachments/assets/5df07ad3-f03e-4e55-b8c2-1ef156f73d1a" />

Additionally, the static margin makes it challenging to left-align the Chart properly with other screen elements.

<img width="646" alt="Screenshot 2025-05-29 at 16 41 33" src="https://github.com/user-attachments/assets/af43e324-9f65-496c-a36e-4c8c05450f40" />


<img width="301" alt="Screenshot 2025-05-29 at 16 25 19" src="https://github.com/user-attachments/assets/0012adb9-eafe-4c05-b6d1-6436d3fbde64" />

This is a common challenge in data visualization. (https://github.com/airbnb/visx/discussions/1086)

The Visx library doesn't provide built-in support for this, but there are some workarounds.

1. Auto-calculate margins - Measure text width and dynamically adjust margins based on data
2. Format/truncate labels - Use scientific notation (1.2B instead of 1,234,567,890) to keep labels short
3. Scale-to-fit approach - Use visx’s scaleToFit feature to shrink long labels within available space automatically (It will wrap the text to the next line)


Following a discussion with the designer (p1748412648081949-slack-C06FSDTSN82), this PR applies both option 1 and option 2 to the line and bar charts. 

The margin is now calculated dynamically, though the calculation may not be exact—it should be sufficient for most cases. We could improve the calculation in the future. Additionally, users can still manually override the margin if necessary.


### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that the context you' working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->



1 Go to charts folder
```sh
cd projects/js-packages/charts/
```

2. Run Storybook app instance
```sh
npm run storybook
```
3. Go to http://localhost:50240/?path=/story/js-packages-charts-types-bar-chart--smart-formatting
4. Confirm that the margin is calculated dynamically and the labels are not cut off.
5. GO to http://localhost:50240/?path=/story/js-packages-charts-types-line-chart--smart-formatting
6. Confirm that the margin is calculated dynamically and the labels are not cut off.

<img width="664" alt="Screenshot 2025-05-29 at 16 34 19" src="https://github.com/user-attachments/assets/eb419d96-e132-4113-a9e9-62cfa37ea4b0" />
<img width="646" alt="Screenshot 2025-05-29 at 16 34 38" src="https://github.com/user-attachments/assets/bdbaaf9e-1535-4e3d-a51f-6e3a6a0adfbc" />

chihsuan committed:
Update package.json and pnpm-lock.yaml to include d3-format version 3.1.0

chihsuan committed:
Add utility functions for Y-axis tick formatting and label width calculation

- Implemented `formatYTick` to format Y-axis ticks based on data value ranges, supporting SI notation for large numbers.
- Added `getLongestLabelWidth` to calculate the width of the longest label in the dataset, considering formatted values.
- Created unit tests for both functions to ensure correct behavior across various edge cases and data scenarios.
- Updated Jest configuration to handle d3-format transformation for testing.

chihsuan committed:
Add Smart Formatting story for LineChart with large values data

- Introduced a new story, `SmartFormatting`, to demonstrate the Smart Formatting feature for Y-axis tick labels in the LineChart component.
- Added a new data file, `large-values-sample.ts`, containing sample data with large values for revenue and users to showcase the formatting capabilities.
- Updated the story's documentation to explain the formatting behavior for different value ranges.

chihsuan committed:
Add useChartMargin hook and refactor Y-axis tick formatting utility

chihsuan committed:
Refactor BarChart and LineChart components to utilize useChartMargin

Dynamic margin calculation and improve Y-axis tick formatting

chihsuan committed:
Add tests

chihsuan committed:
Add changelog

chihsuan committed:
Add story for bar chart

kangzj commented on projects/js-packages/charts/src/components/shared/utils.ts:
```diff
@@ -0,0 +1,69 @@
+import { getStringWidth } from '@visx/text';
+import { format } from 'd3-format';
```
I wonder if we would be able to leverage `@automattic/number-formatters` instead?

chihsuan replied:
That's a good idea. I didn't know we had that package. I'll check it out.

chihsuan replied:
I've removed `d3-format` dependency and use `@automattic/number-formatters` instead. d705d4098d3cd15c4d95024cb459554284241ab5

kangzj commented on projects/js-packages/charts/src/components/shared/utils.ts:
```diff
@@ -0,0 +1,69 @@
+	const maxAbs = Math.max( ...data.map( d => Math.abs( Number( d.value ) || 0 ) ) );
+
+	if ( maxAbs >= 1e9 ) {
+		return format( ',.3s' ); // Billions: 1.23G
```
I'm not expert on this matter. It feels `G` is only relevant for storage or bandwidth, but for numbers in general, at least for the `en_US` locale, they usually use `b`.

> 1,000,000,000 (one [billion](https://en.wikipedia.org/wiki/Billion), [short scale](https://en.wikipedia.org/wiki/Long_and_short_scale); one thousand million or one milliard, one yard,[[1]](https://en.wikipedia.org/wiki/1,000,000,000#cite_note-1) [long scale](https://en.wikipedia.org/wiki/Long_and_short_scales)) is the [natural number](https://en.wikipedia.org/wiki/Natural_number) following [999,999,999](https://en.wikipedia.org/wiki/100,000,000#900,000,000_to_999,999,999) and preceding 1,000,000,001. With a number, "billion" can be abbreviated as b, bil[[2]](https://en.wikipedia.org/wiki/1,000,000,000#cite_note-2) or bn.[[3]](https://en.wikipedia.org/wiki/1,000,000,000#cite_note-3)[[4]](https://en.wikipedia.org/wiki/1,000,000,000#cite_note-EUStyleGuide-4)

Ref: https://en.wikipedia.org/wiki/1,000,000,000

chihsuan replied:
While "b" is typically used to abbreviate billion, the "s" format specifier in d3 refers to SI-prefix formatting, which relies on metric prefixes—so "G" (for giga) is used instead.

But you raised a good point. I think we should use a better format specifier. 🤔

chihsuan replied:
With `@automattic/number-formatters`, it now uses `B` for billion.

<img width="652" alt="Screenshot 2025-06-02 at 15 24 42" src="https://github.com/user-attachments/assets/ca717243-829b-423c-ac6b-7d22b321df04" />
<img width="658" alt="Screenshot 2025-06-02 at 15 24 33" src="https://github.com/user-attachments/assets/50970285-2dc1-4349-b714-e8211bdf65f1" />

I've also enhanced the margin calculation. The `useChartMargin` hook now dynamically determines tick widths, rather than relying on min/max value checks.

kangzj commented on projects/js-packages/charts/src/components/shared/utils.ts:
```diff
@@ -0,0 +1,69 @@
+	const longestLabel =
+		formattedMaxValue.length >= formattedMinValue.length ? formattedMaxValue : formattedMinValue;
+
+	const stringWidth = getStringWidth( longestLabel, labelStyle );
```
👍

kangzj reviewed (commented):
Loving your work @chihsuan ❤️ 

Left two questions, but it looks awesome in general 👍 

This should solve the issue for Jetpack Stats as well: https://github.com/Automattic/wp-calypso/blob/9650e99c366a21383425fa770575de56cd49896f/client/my-sites/stats/components/line-chart/index.tsx#L203

Nice work!

chihsuan committed:
Remove d3-format dependency and add @automattic/number-formatters to package.json and pnpm-lock.yaml

chihsuan committed:
Refactor BarChart and LineChart components to improve axis tick formatting and margin calculation

- Introduced `formatNumberCompact` for Y-axis tick formatting.
- Updated `useChartMargin` to dynamically calculate margins based on chart options and data.
- Refactored axis rendering to utilize new chart options structure for better maintainability.
- Enhanced utility functions for calculating tick widths and formatting.

chihsuan committed:
Update default margin

kangzj commented on projects/js-packages/charts/tests/jest.config.cjs:
```diff
@@ -10,6 +10,8 @@ module.exports = {
 		'<rootDir>/tests/jest.setup.js',
 	],
+	// d3-format is not transpiled to CommonJS, so we need to transform it.
+	transformIgnorePatterns: [ 'node_modules/(?!(?:\\.pnpm|d3-format)/)' ],
```
```suggestion
```
Guess we could remove these now.

chihsuan replied:
Good catch!

kangzj commented on projects/js-packages/charts/src/components/shared/test/use-chart-margin.test.tsx:
```diff
@@ -0,0 +1,119 @@
+			options.axis.y.tickFormat,
+			theme.axisStyles.y.left.axisLabel
+		);
+		expect( result.current.left ).toBe( 48 ); // 40 + 8
```
👍

kangzj reviewed (dismissed):
Works well and looks good!

Might just need to remove the not used config and then merge. Great work 👍

kangzj commented:
Plus we need to fix this JS test:

<img width="820" alt="image" src="https://github.com/user-attachments/assets/5cb7246a-e6ec-4b17-87c1-4811f1fcabb8" />

chihsuan committed:
Update projects/js-packages/charts/tests/jest.config.cjs

Co-authored-by: Jasper Kang <jasper.kang@automattic.com>

chihsuan committed:
Merge branch 'trunk' into charts-2-axis-label-overflow-in-line-and-bar-charts

chihsuan committed:
Fix tests

chihsuan commented:
> Plus we need to fix this JS test:
> 
> <img alt="image" width="820" src="https://private-user-images.githubusercontent.com/1425433/450498261-5cb7246a-e6ec-4b17-87c1-4811f1fcabb8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDg5MTk0NjgsIm5iZiI6MTc0ODkxOTE2OCwicGF0aCI6Ii8xNDI1NDMzLzQ1MDQ5ODI2MS01Y2I3MjQ2YS1lNmVjLTRiMTctODdjMS00ODExZjFmY2FiYjgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDYwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA2MDNUMDI1MjQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OGY4YjU3ZmM3MWQxNTZiYzgxNzFjY2UxNDdmOGU2MTYzYzQ5ZTg3MzAzZjdjMjZkNTdhODBhMDExYjA2ODFlNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.NadxCeiU3fD3lzeYyFBKh0ssNa0TYezsD8WXz9GWbog">

Fixed!
